### PR TITLE
add DUF test code

### PR DESCRIPTION
**Pull request by ckkelvinchan**

### CODE REVIEW

File: codes/models/modules/DUF_arch.py [hunk 1]
```diff
@@ -0,0 +1,682 @@
...
+    def __init__(self, filter_size=(1, 5, 5), device=torch.device('cuda')):
+        super(DynamicUpsamplingFilter_3C, self).__init__()
+        # generate a local expansion filter, used similar to im2col
+        nF = np.prod(filter_size)
+        expand_filter_np = np.reshape(np.eye(nF, nF),
+                                      (nF, filter_size[0], filter_size[1], filter_size[2]))
+        expand_filter = torch.from_numpy(expand_filter_np).float()
+        self.expand_filter = torch.cat((expand_filter, expand_filter, expand_filter),
+                                       0)  # [75, 1, 5, 5]
+        self.device = device
```
Comment (xinntao) [inline codes/models/modules/DUF_arch.py]: remove `self.device = device`

File: codes/models/modules/DUF_arch.py [hunk 2]
```diff
@@ -0,0 +1,682 @@
...
+    def forward(self, x, filters):
+        '''x: input image, [B, 3, H, W]
+        filters: generate dynamic filters, [B, F, R, H, W], e.g., [B, 25, 16, H, W]
+            F: prod of filter kernel size, e.g., 5*5 = 25
+            R: used for upsampling, similar to pixel shuffle, e.g., 4*4 = 16 for x4
+        Return: filtered image, [B, 3*R, H, W]
+        '''
+        B, nF, R, H, W = filters.size()
+        # using group convolution
+        input_expand = F.conv2d(x, self.expand_filter.to(self.device), padding=2,
```
Comment (xinntao) [inline codes/models/modules/DUF_arch.py]: change `self.expand_filter.to(self.device)` to `self.expand_filter.type_as(x)`

File: codes/models/modules/DUF_arch.py [hunk 3]
```diff
@@ -0,0 +1,682 @@
...
+                                groups=3)  # [B, 75, H, W] similar to im2col
+        input_expand = input_expand.view(B, 3, nF, H, W).permute(0, 3, 4, 1, 2)  # [B, H, W, 3, 25]
+        filters = filters.permute(0, 3, 4, 1, 2)  # [B, H, W, 25, 16]
+        out = torch.matmul(input_expand, filters)  # [B, H, W, 3, 16]
+        return out.permute(0, 3, 4, 1, 2).view(B, 3 * R, H, W)  # [B, 3*16, H, W]
+
+
+class DUF_16L(nn.Module):
+    ''' same structure as that in the paper
```
Comment (xinntao) [inline codes/models/modules/DUF_arch.py]: remove these two lines, add meaningful comments, such as 'Official DUF structure with 16 layers.'

File: codes/models/modules/DUF_arch.py [hunk 4]
```diff
@@ -0,0 +1,682 @@
...
+                                   bias=True)
+
+        self.conv3d_f1 = nn.Conv3d(256, 512, (1, 1, 1), stride=(1, 1, 1), padding=(0, 0, 0),
+                                   bias=True)
+        self.conv3d_f2 = nn.Conv3d(512, 1 * 5 * 5 * 16, (1, 1, 1), stride=(1, 1, 1),
+                                   padding=(0, 0, 0), bias=True)
+
+        self.dynamic_filter = DynamicUpsamplingFilter_3C((1, 5, 5))
+
+        # initialization
```
Comment (xinntao) [inline codes/models/modules/DUF_arch.py]: remove the initialization part

File: codes/models/modules/DUF_arch.py [hunk 5]
```diff
@@ -0,0 +1,682 @@
...
+
+        # dynamic filter
+        out = self.dynamic_filter(x_center, Fx)  # [B, 3*R, H, W]
+        out += Rx.squeeze_(2)
+        out = F.pixel_shuffle(out, 4)  # [B, 3, H, W]
+
+        return out
+
+
+class DenseBlock_noBN(nn.Module):
```
Comment (xinntao) [inline codes/models/modules/DUF_arch.py]: this function is no longer used, remove it

File: codes/models/modules/DUF_arch.py [hunk 6]
```diff
@@ -0,0 +1,682 @@
...
+        x3 = self.lrelu(self.conv3d_5(x2))
+        x3 = self.lrelu(self.conv3d_6(x3))
+        if self.t_reduce:
+            x3 = torch.cat((x2[:, :, 1:-1, :, :], x3), 1)
+        else:
+            x3 = torch.cat((x2, x3), 1)
+        return x3
+
+
+class DUF_16L_noBN(nn.Module):
```
Comment (xinntao) [inline codes/models/modules/DUF_arch.py]: remove this class, too

File: codes/models/modules/DUF_arch.py [hunk 7]
```diff
@@ -0,0 +1,682 @@
...
+        x8 = torch.cat((x7, x8), 1)
+
+        x9 = self.conv3d_17(F.relu(self.bn3d_17(x8), inplace=True))
+        x9 = self.conv3d_18(F.relu(self.bn3d_18(x9), inplace=True))
+        x9 = torch.cat((x8, x9), 1)
+        return x9
+
+
+class DUF_28L(nn.Module):
+    ''' same structure as that in the paper
```
Comment (xinntao) [inline codes/models/modules/DUF_arch.py]: modify the comments

File: codes/models/modules/DUF_arch.py [hunk 8]
```diff
@@ -0,0 +1,682 @@
...
+
+        # dynamic filter
+        out = self.dynamic_filter(x_center, Fx)  # [B, 3*R, H, W]
+        out += Rx.squeeze_(2)
+        out = F.pixel_shuffle(out, 4)  # [B, 3, H, W]
+        return out
+
+
+class DenseBlock_28L_A(nn.Module):
+    '''Dense block w/ BN
```
Comment (xinntao) [inline codes/models/modules/DUF_arch.py]: For the class `class DenseBlock_28L_A`:
1. modify the comments to be more meaningful. explain the purpose of this class. why call it xxx_A
2. I like the codes of `52L`. Could it also be possible to change the 28L to the style of `52L`.

File: codes/models/modules/DUF_arch.py [hunk 9]
```diff
@@ -0,0 +1,682 @@
...
+                                   bias=True)
+
+        self.bn3d_17 = nn.BatchNorm3d(nf + 8 * ng, eps=1e-3, momentum=1e-3)
+        self.conv3d_17 = nn.Conv3d(nf + 8 * ng, nf + 8 * ng, (1, 1, 1), stride=(1, 1, 1),
+                                   padding=(0, 0, 0), bias=True)
+        self.bn3d_18 = nn.BatchNorm3d(nf + 8 * ng, eps=1e-3, momentum=1e-3)
+        self.conv3d_18 = nn.Conv3d(nf + 8 * ng, ng, (3, 3, 3), stride=(1, 1, 1), padding=pad,
+                                   bias=True)
+
+        # initialization
```
Comment (xinntao) [inline codes/models/modules/DUF_arch.py]: remove all the initialization part in this file

File: codes/models/modules/DUF_arch.py [hunk 10]
```diff
@@ -0,0 +1,682 @@
...
+class DenseBlock_52L_A(nn.Module):
+    '''Dense block w/ BN
+    for the second denseblock, t_reduced = True'''
+
+    def __init__(self, nf=64, ng=16):
+        super(DenseBlock_52L_A, self).__init__()
+        pad = (1, 1, 1)
+
+        dense_block_l = []
+        for i in range(0, 21):
```
Comment (xinntao) [inline codes/models/modules/DUF_arch.py]: This snippet of the codes is good :+1:

File: codes/models/modules/DUF_arch.py [hunk 11]
```diff
@@ -0,0 +1,682 @@
...
+            dense_block_l.append(
+                nn.Conv3d(nf + i * ng, nf + i * ng, (1, 1, 1), stride=(1, 1, 1), padding=(0, 0, 0),
+                          bias=True))
+
+            dense_block_l.append(nn.BatchNorm3d(nf + i * ng, eps=1e-3, momentum=1e-3))
+            dense_block_l.append(nn.ReLU())
+            dense_block_l.append(
+                nn.Conv3d(nf + i * ng, ng, (3, 3, 3), stride=(1, 1, 1), padding=pad, bias=True))
+
+        # initialization
```
Comment (xinntao) [inline codes/models/modules/DUF_arch.py]: remove it

File: codes/models/modules/DUF_arch.py [hunk 12]
```diff
@@ -0,0 +1,682 @@
...
+        for m in dense_block_l:
+            _initialize_weights(m, 0.1)
+
+        self.dense_blocks = nn.ModuleList(dense_block_l)
+
+    def forward(self, x):
+        for i in range(0, len(self.dense_blocks), 6):
+            y = x.clone()
```
Comment (xinntao) [inline codes/models/modules/DUF_arch.py]: I do not think clone is necessary here. Check whether clone can be moved here.

File: codes/models/modules/DUF_arch.py [hunk 13]
```diff
@@ -0,0 +1,682 @@
...
+
+        # image residual
+        Rx = self.conv3d_r2(F.relu(self.conv3d_r1(x), inplace=True))  # [B, 3*16, 1, H, W]
+
+        # filter
+        Fx = self.conv3d_f2(F.relu(self.conv3d_f1(x), inplace=True))  # [B, 25*16, 1, H, W]
+        Fx = F.softmax(Fx.view(B, 25, 16, H, W), dim=1)
+
+        # Adapt to official model weights
+        if self.adapt_official:
```
Comment (xinntao) [inline codes/models/modules/DUF_arch.py]: This adapt_official part has appeared several times. Can it be separated into a function?

File: codes/utils/util.py [hunk 14]
```diff
@@ -126,6 +127,36 @@ def save_img(img, img_path, mode='RGB'):
     cv2.imwrite(img_path, img)
 
 
+# Gaussian kernel for downsampling used in DUF official code
```
Comment (xinntao) [inline codes/utils/util.py]: In the original DUF repo, it supports x2, x3, and x4. Can we also support these?
And maybe we can also convert the x2, x3 modes, too.

File: codes/models/modules/DUF_arch.py [hunk 15]
```diff
@@ -0,0 +1,682 @@
...
+        self.conv3d_4 = nn.Conv3d(nf + ng, ng, (3, 3, 3), stride=(1, 1, 1), padding=pad, bias=True)
+
+        self.bn3d_5 = nn.BatchNorm3d(nf + 2 * ng, eps=1e-3, momentum=1e-3)
+        self.conv3d_5 = nn.Conv3d(nf + 2 * ng, nf + 2 * ng, (1, 1, 1), stride=(1, 1, 1),
+                                  padding=(0, 0, 0), bias=True)
+        self.bn3d_6 = nn.BatchNorm3d(nf + 2 * ng, eps=1e-3, momentum=1e-3)
+        self.conv3d_6 = nn.Conv3d(nf + 2 * ng, ng, (3, 3, 3), stride=(1, 1, 1), padding=pad,
+                                  bias=True)
+
+        # initialization
```
[the remark this answers is not in the document]
Comment (ckkelvinchan) [reply codes/models/modules/DUF_arch.py]: Do you mean all initializations or only the ones you mentioned here and below?

**Pull request merged**